### PR TITLE
PubSub, Network Shaping & Metrics

### PR DESCRIPTION
**Pull request by SionoiS**
This is the most up to date branch.

### CODE REVIEW

File: src/background.rs [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+    pub async fn new(
+        client_rx: mpsc::UnboundedReceiver<Command>,
+        params: RunParameters,
+    ) -> Result<Self, Box<dyn std::error::Error>> {
+        let (websocket_tx, req_rx) = mpsc::unbounded_channel();
+        let (res_tx, websocket_rx) = mpsc::unbounded_channel();
+
+        let web = WebsocketClient::new(res_tx, req_rx).await?;
+
+        tokio::spawn(web.run());
```
Comment (mxinden) [inline src/background.rs:120]: @[mxinden](https://github.com/mxinden) mxinden [yesterday](https://github.com/testground/sdk-rust/pull/4#discussion_r827974432)

Is spawning the websocket work onto a separate task necessary? If I am not mistaken, Commands like Command::Publish await the flushing of the message by the separate task via the provided oneshot anyways, thus making this a synchronous operation. Would it not be simpler to do all of this in a single task?

I doubt we need to optimize for performance on any of this.

---


@[SionoiS](https://github.com/SionoiS) SionoiS [yesterday](https://github.com/testground/sdk-rust/pull/4#discussion_r828222519)

I failed at combining the 2. It's surely possible but my async knownlege is not there yet.


---

@SionoiS I can propose a patch against this pull request if you like.

Comment (SionoiS) [reply src/background.rs:120]: Can't wait to see what you come up with!

**Comment by mxinden** [comment]
See https://github.com/SionoiS/sdk-rust/pull/2 for a suggestion on folding `websocket.rs` into `background.rs` @SionoiS.

**Comment by ackintosh** [comment]
Great work. I look forward to this PR to be merged. 😃

**Review by mxinden — APPROVED** [review]
Thanks again @SionoiS for the help here. 

I will merge this pull request. There are a couple of changes I would like to propose before cutting a new release, though I think it is a lot easier doing these in separate pull requests.

**Comment by mxinden** [comment]
> Great work. I look forward to this PR to be merged. smiley

Out of curiosity @ackintosh what is your use-case for testground? Are you planning to test lighhouse with it?

File: src/client.rs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+const BACKGROUND_SENDER: &str = "Background Sender";
+
+/// Basic synchronization client enabling one to send signals, await barriers and subscribe or publish to a topic.
+#[derive(Clone)]
+pub struct Client {
+    cmd_tx: UnboundedSender<Command>,
+}
+
+impl Client {
+    pub async fn new() -> Result<(Self, RunParameters), Box<dyn std::error::Error>> {
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply src/client.rs:31]: Your idea would indeed, make more sense then what I came up with.

**Comment by ackintosh** [comment]
@mxinden Yes, I’m planning to test [discv5](https://github.com/sigp/discv5) which is used in lighthouse, with testground.

**Comment by mxinden** [comment]
> @mxinden Yes, I’m planning to test [discv5](https://github.com/sigp/discv5) which is used in lighthouse, with testground.

That is great to hear. Let me know how it goes @ackintosh.

**Comment by ackintosh** [comment]
@mxinden FYI: I've finished initial implementation for the test plan for discv5 [here](https://github.com/ackintosh/test-plan-discv5).

**Comment by mxinden** [comment]
Thanks for sharing @ackintosh. This is really cool! Great to see this in use.

As an aside, I am working on a first compatibility test between go-libp2p and rust-libp2p (https://github.com/libp2p/test-plans/pull/20). Might be interesting for you to test discv5 Rust and Golang compatibility sometime in the future.

**Pull request merged**